### PR TITLE
E0308: return trait implementor

### PR DESCRIPTION
**Pull request by Songbird0**
An example was added to cover the "how to return a trait?" case.

Feel free to correct my orthographic/grammar mistakes (if any) at the same time. I'm not really good.

This PR is linked to #44228.

cc @GuillaumeGomez

### CODE REVIEW

File: src/librustc/diagnostics.rs [hunk 1]
```diff
@@ -1317,6 +1317,55 @@ let x: i32 = "I am not a number!";
...
+```compile_fail,E0308
+trait A {}
+
+struct B;
+
+impl A for B {}
+
+fn foo(b: B) -> A {
+    // will not work
+    b
```
Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Please write "error: [the error message]" in here instead.

File: src/librustc/diagnostics.rs [hunk 2]
```diff
@@ -1317,6 +1317,55 @@ let x: i32 = "I am not a number!";
...
+# use std::boxed::Box;
+# 
+# trait A {}
+#
+# struct B;
+#
+# impl A for B {}
+#
+fn foo(b: B) -> Box<A> {
+    // the return type has been "generified" since the `Box<T>` input parameter is generic.
```
Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Missing capital at the first letter of the first word.

Comment (GuillaumeGomez) [reply src/librustc/diagnostics.rs]: Also, "generified"? Sounds super strange haha.

File: src/librustc/diagnostics.rs [hunk 3]
```diff
@@ -1317,6 +1317,55 @@ let x: i32 = "I am not a number!";
...
+2.
+```rust
+# trait A {}
+#
+# struct B;
+#
+# impl A for B {}
+#
+fn foo<T: A>(b: T) -> T {
+    // generic return type, ok !
```
Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: Missing capital. No whitespace before "!".

**Comment by Songbird0** [comment]
That should be ok. My apologies, my english is `rust`y. :D

File: src/librustc/diagnostics.rs [hunk 4]
```diff
@@ -1317,6 +1317,58 @@ let x: i32 = "I am not a number!";
 //      |
 //    type `i32` assigned to variable `x`
 ```
+
+You will get this error too if you want to return a trait (implementor):
```
Comment (nikomatsakis) [inline src/librustc/diagnostics.rs]: Hmm. I wonder if we should assign this case a distinct error code; it feels like this text is getting awfully long, but also somehow not long enough. In other words, it'd maybe be good to provide a few pointers into the book and give just a *bit* more detail.

Comment (nikomatsakis) [reply src/librustc/diagnostics.rs]: In particular, I'd maybe want to talk about how the actual meaning of this is to return a "trait object", which is likely not what you expected. Not sure just how to approach that topic in this limited space though.

**Review by nikomatsakis — COMMENTED** [review]
Left some thoughts -- I like the idea of giving useful advice on this topic! Have to think about the clearest way to present it, I think.

**Comment by nikomatsakis** [comment]
cc @steveklabnik @carols10cents -- maybe you have thoughts about good reference material to link to? The context is that we are trying to help users who get errors when writing code like `fn foo() -> Iterator { ... }` (or any trait).

**Comment by carols10cents** [comment]
Here's the section in the book on trait objects: https://doc.rust-lang.org/book/second-edition/ch17-02-trait-objects.html

**Comment by nikomatsakis** [comment]
Alternatively, it occurs to me that this is **precisely** the confusion we are hoping to address with the `dyn Trait` RFC. Maybe it'd be better to leave things as they are and press on that direction instead.

**Comment by Songbird0** [comment]
Hi!

> Maybe it'd be better to leave things as they are and press on that direction instead.

That's fine! You may close both #44228 and #44227. :)

**Comment by nikomatsakis** [comment]
Closing, as per last comment. Thanks @Songbird0 in any case!

**Pull request closed without merging**